pkg/providers/v1/variant/fargate/fargate.go

```diff
@@ -0,0 +1,170 @@
+			addresses = append(addresses, nodeAddresses...)
+		case "ipv6":
+			if eni.Ipv6Addresses == nil || len(eni.Ipv6Addresses) == 0 {
+				klog.Errorf("no Ipv6Addresses associated with the eni")
+				continue
+			}
+			internalIPv6Address := eni.Ipv6Addresses[0].Ipv6Address
+			nodeAddresses := getNodeAddressesForFargateNode(awssdk.StringValue(eni.PrivateDnsName), awssdk.StringValue(internalIPv6Address))
+			addresses = append(addresses, nodeAddresses...)
+		}
+	}
+	return addresses, nil
+}
+
+func (f *fargateVariant) InstanceExists(instanceID, vpcID string) (bool, error) {
+	eni, err := f.DescribeNetworkInterfaces(f.ec2API, instanceID, vpcID)
+	return eni != nil, err
+}
+
+func (f *fargateVariant) InstanceShutdown(instanceID, vpcID string) (bool, error) {
+	eni, err := f.DescribeNetworkInterfaces(f.ec2API, instanceID, vpcID)
+	return eni != nil, err
+}
+
+func newEc2Filter(name string, values ...string) *ec2.Filter {
+	filter := &ec2.Filter{
+		Name: awssdk.String(name),
+	}
+	for _, value := range values {
+		filter.Values = append(filter.Values, awssdk.String(value))
+	}
+	return filter
+}
+
+const (
+	// privateDNSNamePrefix is the prefix added to ENI Private DNS Name.
+	privateDNSNamePrefix = "ip-"
+)
+
+// extract private ip address from node name
+func nodeNameToIPAddress(nodeName string) string {
+	nodeName = strings.TrimPrefix(nodeName, privateDNSNamePrefix)
+	nodeName = strings.Split(nodeName, ".")[0]
+	return strings.ReplaceAll(nodeName, "-", ".")
+}
+
+// DescribeNetworkInterfaces returns network interface information for the given DNS name.
+func (f *fargateVariant) DescribeNetworkInterfaces(ec2API iface.EC2, instanceID, vpcID string) (*ec2.NetworkInterface, error) {
+	eniEndpoint := strings.TrimPrefix(instanceID, fargateNodeNamePrefix)
+
+	filters := []*ec2.Filter{
+		newEc2Filter("attachment.status", "attached"),
+		newEc2Filter("vpc-id", vpcID),
+	}
+
+	// when enableDnsSupport is set to false in a VPC, interface will not have private DNS names.
+	// convert node name to ip address because ip-name based and resource-named EC2 resources
+	// may have different privateDNSName formats but same privateIpAddress format
+	if strings.HasPrefix(eniEndpoint, privateDNSNamePrefix) {
+		eniEndpoint = nodeNameToIPAddress(eniEndpoint)
+	}
+
+	filters = append(filters, newEc2Filter("private-ip-address", eniEndpoint))
+
+	request := &ec2.DescribeNetworkInterfacesInput{
+		Filters: filters,
+	}
+
+	eni, err := ec2API.DescribeNetworkInterfaces(request)
+	if err != nil {
+		return nil, err
+	}
+	if len(eni.NetworkInterfaces) == 0 {
+		return nil, nil
+	}
+	if len(eni.NetworkInterfaces) != 1 {
+		// This should not be possible - ids should be unique
+		return nil, fmt.Errorf("multiple interfaces found with same id %q", eni.NetworkInterfaces)
+	}
+	return eni.NetworkInterfaces[0], nil
+}
+
+func init() {
+	v := &fargateVariant{}
+	variant.RegisterVariant(
+		"fargate",
+		v,
+	)
+}
```

Comment:
Something connected to `main()` has to `import` this package for the init to have any effect, right? (maybe I missed it)

We could do this registration in `variant.go` instead, but seems like the goal is to allow other variants to be dropped-in without patching other bits?

Reply:
Ah I missed it: https://github.com/kubernetes/cloud-provider-aws/pull/917/files#diff-97a31d0f193ea0b4e2c538595daf93405351054377458053a5010f6fdb97f4dcR63

I think it makes more sense for that to be in `variants.go`, but I'd ship it

Reply:
we can always iterate. thanks!